helix-core/src/main/java/org/apache/helix/task/TaskDriver.java

```diff
@@ -585,36 +585,195 @@ public void addTask(String workflowName, String jobName, TaskConfig taskConfig,
+    JobConfig jobConfig = getJobConfig(nameSpaceJobName);
+    if (jobConfig == null) {
+      throw new IllegalArgumentException("Job " + nameSpaceJobName + " config does not exist!");
+    }
+
+    TaskConfig taskConfig = null;
+    Map<String, TaskConfig> allTaskConfigs = jobConfig.getTaskConfigMap();
+    for (Map.Entry<String, TaskConfig> entry : allTaskConfigs.entrySet()) {
+      if (entry.getKey().equals(taskID)) {
+        taskConfig = entry.getValue();
+      }
+    }
+
+    validateConfigsForTaskModifications(workflowName, jobName, taskConfig);
+
+    WorkflowContext workflowContext = getWorkflowContext(workflowName);
+    JobContext jobContext = getJobContext(nameSpaceJobName);
+    // If workflow context or job context is null. It means job has not been started. Hence task can
+    // be deleted from the job
+    if (workflowContext != null && jobContext != null) {
+      TaskState jobState = workflowContext.getJobState(nameSpaceJobName);
+      if (jobState != null && ILLEGAL_JOB_STATES_FOR_TASK_MODIFICATION.contains(jobState)) {
+        throw new HelixException("Job " + nameSpaceJobName
+            + " is in illegal state for task deletion. Job State is " + jobState);
+      }
+    }
+
+    DataUpdater<ZNRecord> taskRemover = new DataUpdater<ZNRecord>() {
+      @Override
+      public ZNRecord update(ZNRecord currentData) {
+        if (currentData != null) {
+          Map<String, Map<String, String>> taskMap = currentData.getMapFields();
+          if (taskMap == null) {
+            LOG.warn("Could not update the jobConfig: " + jobName + " Znode MapField is null.");
+            return null;
+          }
+          Map<String, Map<String, String>> newTaskMap = new HashMap<String, Map<String, String>>();
+          for (Map.Entry<String, Map<String, String>> entry : taskMap.entrySet()) {
+            if (!entry.getKey().equals(taskID)) {
+              newTaskMap.put(entry.getKey(), entry.getValue());
+            }
+          }
+          currentData.setMapFields(newTaskMap);
+        }
+        return currentData;
+      }
+    };
+
+    updateTaskInJobConfig(workflowName, jobName, taskRemover);
+
+    workflowContext =
+        _accessor.getProperty(_accessor.keyBuilder().workflowContextZNode(workflowName));
+    jobContext =
+        _accessor.getProperty(_accessor.keyBuilder().jobContextZNode(workflowName, jobName));
+
+    if (workflowContext == null || jobContext == null) {
+      return;
     }
-    addTaskToJobConfig(workflowName, jobName, taskConfig, endTime);
+
+    while (System.currentTimeMillis() <= endTime) {
+      jobContext =
+          _accessor.getProperty(_accessor.keyBuilder().jobContextZNode(workflowName, jobName));
+      workflowContext =
+          _accessor.getProperty(_accessor.keyBuilder().workflowContextZNode(workflowName));
+      if (!jobContext.getTaskIdPartitionMap().containsKey(taskID)
+          && workflowContext.getJobState(nameSpaceJobName) == TaskState.IN_PROGRESS) {
```

Comment:
This condition may not be correct. If this is the only task, job can be completed or something.

Reply:
Yeah, I agree. removed the IN_PROGRESS condition.